File: packages/python/minions_openclaw/gateway_client.py
"""Gateway WebSocket client."""
from __future__ import annotations
import asyncio
import json
import hashlib
import hmac
import base64
from typing import Any, Dict, Optional

try:
    import websockets
    HAS_WEBSOCKETS = True
except ImportError:
    HAS_WEBSOCKETS = False


class GatewayClient:
    def __init__(self, url: str, token: Optional[str] = None, device_private_key: Optional[str] = None) -> None:
        self.url = url
        self.token = token
        self.device_private_key = device_private_key
        self._ws = None
        self._device_token: Optional[str] = None
        self._connected = False

    async def open_connection(self) -> None:
        if not HAS_WEBSOCKETS:
            raise RuntimeError("websockets package not installed. Run: pip install websockets")
        import websockets as ws_lib
        headers = {}
        if self.token:
            headers['Authorization'] = f'Bearer {self.token}'
        self._ws = await ws_lib.connect(self.url, additional_headers=headers)
        msg = json.loads(await asyncio.wait_for(self._ws.recv(), timeout=10))
        if msg.get('type') == 'connect.challenge':
            nonce = msg['payload'].get('nonce', '')
            timestamp = msg['payload'].get('timestamp', '')
            signature = self._sign_challenge(nonce, timestamp)
            await self._ws.send(json.dumps({
                'type': 'connect',
                'payload': {
                    'role': 'operator',
                    'scopes': ['operator.read'],
                    'signature': signature,
                    'timestamp': timestamp,
                    'nonce': nonce,
                    **(({'deviceToken': self._device_token}) if self._device_token else {}),
                }
            }))
            response = json.loads(await asyncio.wait_for(self._ws.recv(), timeout=10))
            if response.get('type') == 'hello-ok':
                self._connected = True
                if response['payload'].get('deviceToken'):
                    self._device_token = response['payload']['deviceToken']
            elif response.get('type') == 'hello-error':
                raise RuntimeError(f"Auth failed: {response.get('payload')}")

    def _sign_challenge(self, nonce: str, timestamp: str) -> str:
        if not self.device_private_key:
            return ''
        try:
            key = self.device_private_key.encode()
            msg_bytes = f'{nonce}:{timestamp}'.encode()
            sig = hmac.new(key, msg_bytes, hashlib.sha256).digest()
            return base64.b64encode(sig).decode()
        except Exception:
            return ''

    async def call(self, method: str, params: Optional[Dict[str, Any]] = None) -> Any:
        if not self._ws:
            raise RuntimeError("Not connected")
        from ._minions_stub import generate_id
        call_id = generate_id()
        await self._ws.send(json.dumps({'type': 'call', 'id': call_id, 'method': method, 'params': params or {}}))
        while True:
            raw = await asyncio.wait_for(self._ws.recv(), timeout=10)
            msg = json.loads(raw)
            if msg.get('id') == call_id:
                return msg.get('payload')

    async def fetch_presence(self) -> Dict[str, Any]:
        results = await asyncio.gather(
            self.call('agents.list'),
            self.call('channels.list'),
            self.call('models.list'),
            self.call('system-presence'),
            return_exceptions=True
        )
        agents_r, channels_r, models_r, config_r = results
        return {
            'agents': (agents_r.get('items', []) if isinstance(agents_r, dict) else []),
            'channels': (channels_r.get('items', []) if isinstance(channels_r, dict) else []),
            'models': (models_r.get('items', []) if isinstance(models_r, dict) else []),
            'config': (config_r if isinstance(config_r, dict) else {}),
        }

    async def close(self) -> None:
        if self._ws:
            await self._ws.close()
            self._ws = None
            self._connected = False

    @property
    def device_token(self) -> Optional[str]:
        return self._device_token

File: packages/core/src/GatewayClient.ts
import WebSocket from 'ws';
import { createSign } from 'crypto';
import { generateId } from 'minions-sdk';

export interface GatewayPresence {
  agents: unknown[];
  channels: unknown[];
  models: unknown[];
  config: Record<string, unknown>;
}

interface GatewayMessage {
  type: string;
  payload: Record<string, unknown>;
  id?: string;
}

export class GatewayClient {
  private ws: WebSocket | null = null;
  private deviceToken: string | null = null;
  private connected = false;

  constructor(
    private readonly url: string,
    private readonly token?: string,
    private readonly devicePrivateKey?: string
  ) {}

  async openConnection(): Promise<void> {
    return new Promise((resolve, reject) => {
      try {
        this.ws = new WebSocket(this.url, {
          headers: this.token ? { Authorization: `Bearer ${this.token}` } : {},
        });

        const timeout = setTimeout(() => {
          reject(new Error('Connection timeout'));
          this.ws?.close();
        }, 10000);

        this.ws.on('message', async (data: WebSocket.RawData) => {
          try {
            const msg = JSON.parse(data.toString()) as GatewayMessage;
            await this.handleMessage(
              msg,
              () => { clearTimeout(timeout); resolve(); },
              (err) => { clearTimeout(timeout); reject(err); }
            );
          } catch (err) {
            clearTimeout(timeout);
            reject(err);
          }
        });

        this.ws.on('error', (err) => {
          clearTimeout(timeout);
          reject(err);
        });

        this.ws.on('close', () => {
          if (!this.connected) {
            clearTimeout(timeout);
            resolve();
          }
        });

      } catch (err) {
        reject(err);
      }
    });
  }

  private async handleMessage(
    msg: GatewayMessage,
    resolve: () => void,
    reject: (err: Error) => void
  ): Promise<void> {
    if (msg.type === 'connect.challenge') {
      const nonce = msg.payload['nonce'] as string;
      const timestamp = msg.payload['timestamp'] as string;
      const signature = this.signChallenge(nonce, timestamp);
      this.send({
        type: 'connect',
        payload: {
          role: 'operator',
          scopes: ['operator.read'],
          signature,
          timestamp,
          nonce,
          ...(this.deviceToken ? { deviceToken: this.deviceToken } : {}),
        },
      });
    } else if (msg.type === 'hello-ok') {
      this.connected = true;
      if (msg.payload['deviceToken']) {
        this.deviceToken = msg.payload['deviceToken'] as string;
      }
      resolve();
    } else if (msg.type === 'hello-error') {
      reject(new Error(`Auth failed: ${JSON.stringify(msg.payload)}`));
    }
  }

  /**
   * Sign the challenge nonce using RSA PKCS1v15 with SHA-256.
   * PKCS1v15 is used here for broad compatibility with existing gateway
   * implementations; matches the Python gateway_client.py behaviour.
   */
  private signChallenge(nonce: string, timestamp: string): string {
    if (!this.devicePrivateKey) return '';
    try {
      const sign = createSign('SHA256');
      sign.update(`${nonce}:${timestamp}`);
      return sign.sign(this.devicePrivateKey, 'base64');
    } catch {
      return '';
    }
  }

  send(msg: Omit<GatewayMessage, 'id'> & { id?: string }): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      throw new Error('WebSocket not open');
    }
    this.ws.send(JSON.stringify(msg));
  }

  async call(method: string, params: Record<string, unknown> = {}): Promise<unknown> {
    return new Promise((resolve, reject) => {
      if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
        reject(new Error('Not connected'));
        return;
      }
      const id = generateId();
      const timeout = setTimeout(() => reject(new Error(`Timeout calling ${method}`)), 10000);

      const handler = (data: WebSocket.RawData) => {
        const msg = JSON.parse(data.toString()) as GatewayMessage;
        if (msg.id === id) {
          clearTimeout(timeout);
          this.ws?.off('message', handler);
          resolve(msg.payload);
        }
      };
      this.ws.on('message', handler);
      this.ws.send(JSON.stringify({ type: 'call', id, method, params }));
    });
  }

  async fetchPresence(): Promise<GatewayPresence> {
    const [agentsRaw, channelsRaw, modelsRaw, configRaw] = await Promise.allSettled([
      this.call('agents.list'),
      this.call('channels.list'),
      this.call('models.list'),
      this.call('system-presence'),
    ]);

    return {
      agents: (agentsRaw.status === 'fulfilled' ? (agentsRaw.value as { items?: unknown[] })?.items ?? [] : []),
      channels: (channelsRaw.status === 'fulfilled' ? (channelsRaw.value as { items?: unknown[] })?.items ?? [] : []),
      models: (modelsRaw.status === 'fulfilled' ? (modelsRaw.value as { items?: unknown[] })?.items ?? [] : []),
      config: (configRaw.status === 'fulfilled' ? configRaw.value as Record<string, unknown> : {}),
    };
  }

  async close(): Promise<void> {
    return new Promise((resolve) => {
      if (!this.ws || this.ws.readyState === WebSocket.CLOSED) {
        resolve();
        return;
      }
      this.ws.once('close', () => resolve());
      this.ws.close();
    });
  }

  getDeviceToken(): string | null {
    return this.deviceToken;
  }
}
